### PR TITLE
Multi prison

### PR DESCRIPTION
**Pull request by alan**
Only to be merged once users have been setup!

### CODE REVIEW

File: app/models/signon_identity.rb [hunk 1]
```diff
@@ -34,21 +32,16 @@ def from_session_data(data)
...
       orgs = permissions.map { |p| p.fetch('organisation') }
 
-      return [] if orgs.empty?
-
-      if orgs.include?(DIGITAL_ORG)
-        Estate.all
-      else
-        Estate.where(sso_organisation_name: orgs)
-      end
+      mapper = EstateSSOMapper.new(orgs)
```
Comment (StupidCodeFactory) [inline app/models/signon_identity.rb]: 👍

File: app/services/estate_sso_mapper.rb [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+      grouped_estates = Hash.new { |h, k| h[k] = [] }
+      Estate.where.not(group: nil).each do |estate|
+        grouped_estates[estate.group] << estate.sso_organisation_name
+      end
+      grouped_estates.values.freeze
+      grouped_estates.freeze
+    end
+  end
+
+  def self.reset_grouped_estates
```
Comment (StupidCodeFactory) [inline app/services/estate_sso_mapper.rb]: I'm not keen on mixing test and production code, this could easily be stubbed/mocked in tests

File: app/services/estate_visit_query.rb [hunk 3]
```diff
@@ -35,7 +31,23 @@ def processed(limit:, prisoner_number: nil)
...
     Visit.
       ready_for_processing.
-      from_estate(@estate).
+      from_estates(@estates).
       count
   end
+
+private
+
+  def grouped_visits(visits)
```
Comment (StupidCodeFactory) [inline app/services/estate_visit_query.rb]: :scream: could we make this readable, i'm thinking you can get this same result using a postgres window function? I know that the intent here is to reuse the the  visits in memory array but this is really hard to read.

File: app/views/prison/dashboards/_prison_switcher.html.erb [hunk 4]
```diff
@@ -1,11 +1,19 @@
...
-      <%= label_tag "estate_id", class: "form-label" do -%>
-        Change prison
-      <% end %>
-      <%= select_tag 'estate_id', options_from_collection_for_select(sso_identity.accessible_estates, 'id', 'name', current_estate.id), class: 'form-control form-control-auto-width' %>
-      <%= submit_tag t('.switch_estate'), class: 'button-link' %>
+<div class="grid-row push-bottom">
+  <div class="column-full">
+    <% content_for :prison_switcher do %>
+      <% if sso_identity.accessible_estates.size > 1 -%>
+        <%= form_tag prison_switch_estates_path, enforce_utf8: false, class: 'form prison-switcher-form' do -%>
```
Comment (StupidCodeFactory) [inline app/views/prison/dashboards/_prison_switcher.html.erb]: out of curiosity why is the enforce utf8 set to false?

File: app/controllers/prison/dashboards_controller.rb [hunk 5]
```diff
@@ -31,44 +30,53 @@ def print_visits
...
+                     prisoner_number: prisoner_number)
   end
 
-  def switch_estate
-    estate = Estate.find(params[:estate_id])
-
-    if sso_identity.accessible_estate?(estate)
-      @_current_estate         = nil
-      session[:current_estate] = estate.id
+  def switch_estates
```
Comment (StupidCodeFactory) [inline app/controllers/prison/dashboards_controller.rb]: Could we have a ```SwitchEstatesController``` and not bundle a tone of unrelated stuff together?

Comment (StupidCodeFactory) [inline app/controllers/prison/dashboards_controller.rb]: possibly sharing common code via a module or a new class (which i'd prefer)

File: config/routes.rb [hunk 6]
```diff
@@ -53,7 +53,7 @@
       get '/processed', action: :processed, as: 'processed_visits'
       get '/print_visits', action: :print_visits, as: 'print_visits'
       get '/search', action: :search, as: :search
-      post '/switch_estate', action: :switch_estate, as: 'switch_estate'
+      post '/switch_estates', action: :switch_estates, as: 'switch_estates'
```
Comment (StupidCodeFactory) [inline config/routes.rb]: ```SwitchEstatesController#create```

```ruby
resource :switch_estates, only: :create
```

File: spec/controllers/prison/feedbacks_controller_spec.rb [hunk 7]
```diff
@@ -8,7 +8,9 @@
 
     before do
       login_user(
-        FactoryGirl.create(:user), FactoryGirl.create(:estate))
+        FactoryGirl.create(:user),
```
Comment (StupidCodeFactory) [inline spec/controllers/prison/feedbacks_controller_spec.rb]: Maybe we can drop the ```FactoryGirl``` would be less cluttered?

File: config/routes.rb (hunk 6 above)
Comment (StupidCodeFactory) [inline config/routes.rb]: Using ```resources``` even if it means a having a controller with only one action isn't a big deal and encourages separation of concerns. I'm weary to see fatter controller than what we have now.

File: app/views/prison/dashboards/_prison_switcher.html.erb (hunk 4 above)
Comment (alan) [inline app/views/prison/dashboards/_prison_switcher.html.erb]: Can't remember why we added it...

File: app/services/estate_sso_mapper.rb (hunk 2 above)
Comment (alan) [inline app/services/estate_sso_mapper.rb]: Yeah, I prefer to avoid generally. The issue here is that the order of the tests matter, so if this spec file runs before the feature specs, then some of those fail, so the best thing is to always cleanup after using the method

File: app/services/estate_visit_query.rb (hunk 3 above)
Comment (alan) [inline app/services/estate_visit_query.rb]: We've looked into it and it is not possible to simplify with window functions

File: app/controllers/prison/switch_estates_controller.rb [hunk 8]
```diff
@@ -0,0 +1,26 @@
+class Prison::SwitchEstatesController < ApplicationController
+  before_action :authorize_prison_request
+  before_action :authenticate_user
+  before_action :verify_switch_estates
+
+  def create
+    estates = Estate.where(id: params[:estates_id])
```
Comment (StupidCodeFactory) [inline app/controllers/prison/switch_estates_controller.rb]: typo? ```estate_ids```

File: app/views/prison/dashboards/search.html.erb [hunk 9]
```diff
@@ -16,3 +16,8 @@
 
 <h3 class="heading-medium"><%= t('.processed') %></h3>
 <%= render 'processed' %>
+<script>
+  var pvb = {
```
Comment (StupidCodeFactory) [inline app/views/prison/dashboards/search.html.erb]: @chrishanes how do you feel about having this as a data attributes on the body element (or the another relevant element) rather than poluting the gloabal JS namespace?
